sympy/matrices/common.py

```diff
@@ -2218,37 +2223,46 @@ def __neg__(self):
         return self._eval_scalar_mul(-1)
 
     @call_highest_priority('__rpow__')
-    def __pow__(self, num):
+    def __pow__(self, exp):
         if self.rows != self.cols:
             raise NonSquareMatrixError()
         a = self
-        jordan_pow = getattr(a, '_matrix_pow_by_jordan_blocks', None)
-        num = sympify(num)
-        if num.is_Number and num % 1 == 0:
+        jordan_pow = getattr (a, '_matrix_pow_by_jordan_blocks', None)
+        exp = sympify(exp)
+
+        if exp.is_zero:
+            return a._new(a.rows, a.cols, lambda i, j: int(i == j))
+        if exp == 1:
+            return a
+
+        if exp.is_Number and exp % 1 == 0:
             if a.rows == 1:
-                return a._new([[a[0]**num]])
-            if num == 0:
-                return self._new(self.rows, self.cols, lambda i, j: int(i == j))
-            if num < 0:
-                num = -num
+                return a._new([[a[0]**exp]])
+            if exp < 0:
+                exp = -exp
                 a = a.inv()
             # When certain conditions are met,
             # Jordan block algorithm is faster than
             # computation by recursion.
-            elif a.rows == 2 and num > 100000 and jordan_pow is not None:
+            elif a.rows == 2 and exp > 100000 and jordan_pow is not None:
                 try:
-                    return jordan_pow(num)
+                    return jordan_pow(exp)
                 except MatrixError:
                     pass
-            return a._eval_pow_by_recursion(num)
-        elif not num.is_Number and num.is_negative is None and a.det() == 0:
-            from sympy.matrices.expressions import MatPow
-            return MatPow(a, num)
-        elif isinstance(num, (Expr, float)):
-            return jordan_pow(num)
-        else:
-            raise TypeError(
-                "Only SymPy expressions or integers are supported as exponent for matrices")
+            return a._eval_pow_by_recursion(exp)
+
+        if a.is_diagonal ():
+            return a._new(a.rows, a.cols, lambda i, j: a[i,j]**exp if i == j else 0)
+
+        if jordan_pow:
+            try:
+                return jordan_pow(exp)
+            except ValueError:
+                if exp.is_negative or exp.is_number:
```

Comment:
I still think this should be catching NonInvertibleMatrixError.

What is the `exp.is_number` check doing?

Maybe it should be `if exp.is_nonnegative is False` (raise if the matrix is non-invertible and the exponent is known to *not* be non-negative.

Reply:
NonInvertibleMatrixError is a MatrixError and a ValueError.

Reply:
exp.is_number to see if attempt was made with a concrete number of the negative persuasion, in this case fail, if we do not know that is concrete number and do not know that is negative return symbolic form.

Reply:
The point is we don't want to catch other ValueErrors.

Reply:
Can you give an example where the `is_number` check is useful here?

Reply:
This is what makes a non-invertible matrix to a non-negative non-integer number fail instead of returning a MatPow.

Reply:
So I guess we want to raise if the matrix is non-invertible (hence catching that here) but then only if the exponent is known not to be a non-negative integer. This check currently fails for something like
```python
Matrix([[0, 1], [0, 0]]) ** Integral(1, (x, 2, 3))
# Compare with:
Matrix([[0, 1], [0, 0]]) ** Integral(1, (x, 2, 3)).doit()
```
Probably the condition we want to check here is
```python
            try:
                return jordan_pow(exp)
            except NonInvertibleMatrixError:
                # Raised by jordan_pow unless exp is definitely known to be a non-negative integer.
                # Here we raise if n is definitely not a non-negative integer but otherwise we can leave
                # this as an unevaluated MatPow.
                if n.is_integer is False or n.is_nonnegative is False:
                    raise
```
which is similar to the check in `_matrix_pow_by_jordan_blocks` except more conservative.

Reply:
Ok for the conservative check but I left the exception as a ValueError since MatrixError can also arise from something unexpected happening during computation which should be signalled to the user.

Reply:
I don't understand what you mean. I am suggesting to catch fewer exceptions here - in particular this should only catch the exact exception type that is expected to be raised. Then all other exceptions *will* be signalled to the user.

Reply:
Ah good point, my brain was backwards was thinking that would catch all MatrixErrors...